website/addons/dataverse/views.py

```diff
@@ -0,0 +1,332 @@
+from framework.exceptions import HTTPError
+
+from website.addons.base import generic_views
+from website.addons.dataverse import client
+from website.addons.dataverse.model import DataverseProvider
+from website.addons.dataverse.settings import DEFAULT_HOSTS
+from website.addons.dataverse.serializer import DataverseSerializer
+from website.oauth.models import ExternalAccount
+from website.project.decorators import (
+    must_have_addon, must_be_addon_authorizer,
+    must_have_permission, must_not_be_registration,
+    must_be_contributor_or_public
+)
+from website.util import rubeus, api_url_for
+from website.util.sanitize import assert_clean
+
+SHORT_NAME = 'dataverse'
+FULL_NAME = 'Dataverse'
+
+dataverse_account_list = generic_views.account_list(
+    SHORT_NAME,
+    DataverseSerializer
+)
+
+dataverse_import_auth = generic_views.import_auth(
+    SHORT_NAME,
+    DataverseSerializer
+)
+
+dataverse_deauthorize_node = generic_views.deauthorize_node(
+    SHORT_NAME
+)
+
+dataverse_get_config = generic_views.get_config(
+    SHORT_NAME,
+    DataverseSerializer
+)
+
+## Auth ##
+
+@must_be_logged_in
+def dataverse_user_config_get(auth, **kwargs):
+    """View for getting a JSON representation of the logged-in user's
+    Dataverse user settings.
+    """
+
+    user_addon = auth.user.get_addon('dataverse')
+    user_has_auth = False
+    if user_addon:
+        user_has_auth = user_addon.has_auth
+
+    return {
+        'result': {
+            'userHasAuth': user_has_auth,
+            'urls': {
+                'create': api_url_for('dataverse_add_user_account'),
+                'accounts': api_url_for('dataverse_account_list'),
+            },
+            'hosts': DEFAULT_HOSTS,
+        },
+    }, http.OK
+
+
+## Config ##
+
+@must_be_logged_in
+def dataverse_add_user_account(auth, **kwargs):
+    """Verifies new external account credentials and adds to user's list"""
+    user = auth.user
+    provider = DataverseProvider()
+
+    host = request.json.get('host').rstrip('/')
+    api_token = request.json.get('api_token')
+
+    # Verify that credentials are valid
+    client.connect_or_error(host, api_token)
+
+    # Note: `DataverseSerializer` expects display_name to be a URL
+    try:
+        provider.account = ExternalAccount(
+            provider=provider.short_name,
+            provider_name=provider.name,
+            display_name=host,       # no username; show host
+            oauth_key=host,          # hijacked; now host
+            oauth_secret=api_token,  # hijacked; now api_token
+            provider_id=api_token,   # Change to username if Dataverse allows
+        )
+        provider.account.save()
+    except KeyExistsException:
+        # ... or get the old one
+        provider.account = ExternalAccount.find_one(
+            Q('provider', 'eq', provider.short_name) &
+            Q('provider_id', 'eq', api_token)
+        )
+        assert provider.account is not None
```

Comment:
This line is unnecessary here. `find_one` will raise an exception if it finds nothing